[Security Solution] Add missing alerts (signals) API endpoints OpenAPI specs

maximpn opened the pull request:
**Addresses:** https://github.com/elastic/kibana/issues/183661

## Summary

This PR adds missing OpenAPI specs for the following API endpoints (Alerts API and Alerts Migration API) available in both Serverless and ESS

- `POST /api/detection_engine/signals/status`
- `POST /api/detection_engine/signals/tags`
- `POST /api/detection_engine/signals/search`

and API endpoints available only in ESS

- `POST /api/detection_engine/signals/migration_status`
- `POST /api/detection_engine/signals/migration`
- `POST /api/detection_engine/signals/finalize_migration`
- `DELETE /api/detection_engine/signals/migration`

maximpn committed:
add missing alerts API OAS

maximpn committed:
add alerts migration API OAS

maximpn committed:
refine added OAS specs

maximpn committed:
add serverless and ess labels

maximpn commented:
Hey @marcelhallmann,

Could you check added OpenAPI specs for correctness? Don't pay too much attention to summaries and descriptions since it will be further refined in https://github.com/elastic/kibana/issues/183702. I used some information from the [documentation page](https://www.elastic.co/guide/en/security/current/signals-api-overview.html#_apply_alert_tags).

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals/set_signal_status/set_signals_status_route.gen.ts:
```diff
@@ -0,0 +1,42 @@
+    conflicts: z.enum(['abort', 'proceed']).optional().default('abort'),
+  })
+  .merge(
+    z.union([
```
There's a type error here, can't merge a zod union with another object

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/create_signals_migration/create_signals_migration.gen.ts:
```diff
@@ -0,0 +1,34 @@
+});
+
+export type CreateAlertsMigrationResponse = z.infer<typeof CreateAlertsMigrationResponse>;
+export const CreateAlertsMigrationResponse = z.object({
```
The migration response can have an error as part of the `ok` response: https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/create_signals_migration_route.ts#L142
Looks like this response should be a union of the error and regular responses

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/delete_signals_migration/delete_signals_migration.gen.ts:
```diff
@@ -0,0 +1,41 @@
+
+export type CleanupAlertsMigrationRequestBody = z.infer<typeof CleanupAlertsMigrationRequestBody>;
+export const CleanupAlertsMigrationRequestBody = z.object({
+  migration_ids: z.array(z.string()).min(1).optional(),
```
This should be required

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/delete_signals_migration/delete_signals_migration.gen.ts:
```diff
@@ -0,0 +1,41 @@
+  updated: z.string().datetime(),
+  error: z
+    .object({
+      message: z.string().optional(),
+      status_code: z.number().int().optional(),
```
`message` and `status_code` within `error` are not optional, though `error` is optional

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/finalize_signals_migration/finalize_signals_migration.gen.ts:
```diff
@@ -0,0 +1,42 @@
+
+export type FinalizeAlertsMigrationRequestBody = z.infer<typeof FinalizeAlertsMigrationRequestBody>;
+export const FinalizeAlertsMigrationRequestBody = z.object({
+  migration_ids: z.array(z.string()).min(1).optional(),
```
not optional

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/finalize_signals_migration/finalize_signals_migration.gen.ts:
```diff
@@ -0,0 +1,42 @@
+  updated: z.string().datetime(),
+  error: z
+    .object({
+      message: z.string().optional(),
+      status_code: z.number().int().optional(),
```
`message` and `status_code` within error are not optional, though error is optional

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/get_signals_migration_status/get_signals_migration_status.schema.yaml:
```diff
@@ -0,0 +1,95 @@
+      tags:
+        - Alerts migration API
+      parameters:
+        - name: from
```
I don't see the codegen for this parameter?

maximpn replied:
It's interesting when `x-codegen-enabled` is omitted then code generation works but doesn't generate all the entities. Since a valid spec may produce invalid code I explicitly added `x-codegen-enabled: true`.

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/get_signals_migration_status/get_signals_migration_status.gen.ts:
```diff
@@ -0,0 +1,47 @@
+
+export type IndexMigrationStatus = z.infer<typeof IndexMigrationStatus>;
+export const IndexMigrationStatus = z.object({
+  index: NonEmptyString.optional(),
+  version: z.number().int().optional(),
+  signal_versions: z.array(AlertVersion).optional(),
+  migrations: z.array(MigrationStatus).optional(),
+  is_outdated: z.boolean().optional(),
```
Every field is required

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/get_signals_migration_status/get_signals_migration_status.gen.ts:
```diff
@@ -0,0 +1,47 @@
+
+export type GetAlertsMigrationStatusResponse = z.infer<typeof GetAlertsMigrationStatusResponse>;
+export const GetAlertsMigrationStatusResponse = z.object({
+  indices: z.array(IndexMigrationStatus).optional(),
```
`indices` is required

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/get_signals_migration_status/get_signals_migration_status.gen.ts:
```diff
@@ -0,0 +1,47 @@
+
+export type AlertVersion = z.infer<typeof AlertVersion>;
+export const AlertVersion = z.object({
+  version: z.number().int().optional(),
+  count: z.number().int().optional(),
```
These are required

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/get_signals_migration_status/get_signals_migration_status.gen.ts:
```diff
@@ -0,0 +1,47 @@
+
+export type MigrationStatus = z.infer<typeof MigrationStatus>;
+export const MigrationStatus = z.object({
+  id: NonEmptyString.optional(),
+  status: z.enum(['success', 'failure', 'pending']).optional(),
+  version: z.number().int().optional(),
+  updated: z.string().datetime().optional(),
```
All required

maximpn committed:
rename GenericErrorResponse to PlatformErrorResponse

maximpn committed:
add siemErrorResponse schema

maximpn committed:
fix SetAlertsStatusRequestBody

maximpn committed:
fix /api/detection_engine/signals/migration response

maximpn committed:
fix /api/detection_engine/signals/migration schema

maximpn committed:
fix /api/detection_engine/signals/finalize_migration schema

maximpn committed:
fix /api/detection_engine/signals/migration_status schema

maximpn committed:
make code generation produced correct zod schemas and types

maximpn committed:
extend error response schemas

maximpn commented:
@marshallmain thank you for the review!

I addressed your comments and also found a few things to fix. Explicitly enabled code generation via `x-codegen-enabled: true` forces code generator to produce request and response Zod schemas. I made it enabled in all added OAS files to verify generated code correctness. A valid OAS may produce an invalid TS file.

maximpn committed:
fix 500 error response schema description

maximpn committed:
add explicit code generation flag

marshallmain commented on x-pack/plugins/security_solution/common/api/detection_engine/signals_migration/create_signals_migration/create_signals_migration.gen.ts:
```diff
@@ -0,0 +1,58 @@
+export type AlertsIndexMigrationSuccess = z.infer<typeof AlertsIndexMigrationSuccess>;
+export const AlertsIndexMigrationSuccess = z.object({
+  index: z.string(),
+  migration_id: z.string(),
+  migration_index: z.string(),
```
To match the API route behavior these should exist but be `null` for the error and skipped migration cases.

maximpn replied:
@marshallmain,

OAS doesn't provide a way to specify explicit `null` fields. There is a `nullable` modifier which marks a type as nullable. In this particular case it's possible to declare a single response type instead of `AlertsIndexMigrationSuccess`, `AlertsIndexMigrationError` and `SkippedAlertsIndexMigration` with `migration_id` and `migration_index` being string or null. Such schema won't describe the response precisely neither.

What's the approach you expect here?

marshallmain replied:
We discussed on slack and the approach is ok as is. Better to leave the properties out of the response schema than have them typed incorrectly or make the rest of the schema less precise.

marshallmain reviewed (commented):
Looks much better, thanks for addressing the comments. I only see one remaining issue.

maximpn commented:
/ci

maximpn committed:
apply generated from OAS TS types

maximpn commented:
/ci

marshallmain reviewed (approved):
👍

maximpn commented:
/ci

banderror reviewed (approved):
Rule Management changes LGTM 👍 

I did a high-level review of the whole diff and checked the changes owned by our team.

Can we please add the new folders to the CODEOWNERS file in a separate PR?
